File: workstations.py
from sys import stdin,stdout
from heapq import heappop, heappush
# Python's implementation of PQ is done using a min heap
n,m = map(int,stdin.readline().split())
arrivals,departures,saves = [],[],0
for _ in range(n):
    a,s = map(int,stdin.readline().split())
    heappush(arrivals,a)
    heappush(departures,a+s)

while arrivals:
    time = heappop(arrivals)
    #departures[0] represents smallest element in min heap
    while time - departures[0] > m: #no saves made since the computer will have to be manually unlocked anyways
        heappop(departures)
    # if a resarcher arrives just after another one has left, a save is made
    if departures[0] <= time:
        heappop(departures)
        saves += 1
        
stdout.write(f'{saves}\n')
    
    

File: workstations.java
import java.io.*;
import java.util.*;

public class workstations {
    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        PrintWriter pw = new PrintWriter(new BufferedWriter(new OutputStreamWriter(System.out)));
        String[] firstLine = br.readLine().split(" ");
        int n = Integer.parseInt(firstLine[0]), m = Integer.parseInt(firstLine[1]), saves = 0;
        // One array for arrival timings, and one PQ for departure timings
        PriorityQueue<Integer> departures = new PriorityQueue<Integer>();
        int[] arrivals = new int[n];
        for (int i = 0; i < n; i++){
            String[] line = br.readLine().split(" ");
            int a = Integer.parseInt(line[0]), s = Integer.parseInt(line[1]);
            arrivals[i] = a;
            departures.add(a+s);
        }
        Arrays.sort(arrivals); // sort arrivals in ascending order
        for (int j = 0; j < n; j++){
            // if earliest arrival time is more than earliest departure time by m minutes, the computer will be locked
            // no saves are made, and a manual unlock will have to be done
            while (arrivals[j] - departures.peek() > m){departures.remove();}
            // if a researcher arrives after another one has left, and before m minutes has elapsed, a save is made
            if (arrivals[j]>= departures.peek()){
                departures.remove();
                saves += 1;
            }
        }
        pw.println(saves);
        br.close();
        pw.close();
    }
}
